frontend/components/tree/js/tree.service.ts

```diff
@@ -1015,43 +1022,121 @@ export class TreeService {
 		this.ViewerService.clearHighlights();
 		this.DocsService.closeDocs();
 
-		if (this.currentSelectedNodes) {
-			for (let i = 0; i < this.currentSelectedNodes.length; i++) {
-				this.currentSelectedNodes[i].selected = false;
+		if (this.currentSelectedNodes.length) {
+			while (this.currentSelectedNodes.length) {
+				const currentNode = this.currentSelectedNodes.pop();
+				currentNode.selected = this.SELECTION_STATES.unselected;
+				this.setSelectionOnParentNodes(currentNode, this.SELECTION_STATES.unselected);
+				if (currentNode.children && currentNode.children.length) {
+					this.currentSelectedNodes = this.currentSelectedNodes.concat(currentNode.children);
+				}
 			}
 		}
 
 		this.currentSelectedNodes = [];
 	}
 
+	/**
+	 * Set the selection state of a nodes parents to represent it's children
+	 * @param nodes the nodes to set a new selection state on
+	 * @param select wether the node should be selected or not
+	 */
+	public traverseNodesAndSetSelected(nodes: any[], select: boolean) {
+
+		nodes = nodes.concat(); // make a copy
+		const setParentNodes = nodes.concat();
+		let currentNode;
+
+		while (nodes.length) {
+			currentNode = nodes.pop();
+
+			const nodeIndex = this.currentSelectedNodes.indexOf(currentNode);
+			const notParentOfUnselected = currentNode.selected !== this.SELECTION_STATES.parentOfUnselected;
+
+			if (select && notParentOfUnselected && nodeIndex === -1) {
+				this.currentSelectedNodes.push(currentNode);
+			} else if (nodeIndex > -1) {
+				this.currentSelectedNodes.splice(nodeIndex, 1);
+			}
+
+			if (currentNode.toggleState !== "invisible") {
+
+				if (currentNode.toggleState !== "parentOfInvisible" && select) {
```

Comment:
again, rejig it and you can avoid checking select multiple times

```` js
if (!select) {
	currentNode.selected = this.SELECTION_STATES.unselected;
} else if(currentNode.toggleState !== "parentOfInvisible") {
    currentNode.selected = this.SELECTION_STATES.parentOfUnselected;
} else {
    currentNode.selected = this.SELECTION_STATES.selected;
}
````

Reply:
Do you mean:

```javascript
if (!select) {
	currentNode.selected = this.SELECTION_STATES.unselected;
} else if(currentNode.toggleState === "parentOfInvisible") {
    currentNode.selected = this.SELECTION_STATES.selected;
} else {
    currentNode.selected = this.SELECTION_STATES.parentOfUnselected;
}
```
Otherwise the logic is different